keps/sig-storage/2639-secret-configmap-protection/README.md

```diff
@@ -0,0 +1,601 @@
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [User Stories](#user-stories)
+    - [Story 1](#story-1)
+    - [Story 2](#story-2)
+    - [Story 3](#story-3)
+  - [Notes/Constraints/Caveats (Optional)](#notesconstraintscaveats-optional)
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Design Details](#design-details)
+  - [Test Plan](#test-plan)
+  - [Graduation Criteria](#graduation-criteria)
+    - [Alpha -&gt; Beta Graduation](#alpha---beta-graduation)
+    - [Beta -&gt; GA Graduation](#beta---ga-graduation)
+    - [Removing a Deprecated Flag](#removing-a-deprecated-flag)
+  - [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+  - [Version Skew Strategy](#version-skew-strategy)
+- [Production Readiness Review Questionnaire](#production-readiness-review-questionnaire)
+  - [Feature Enablement and Rollback](#feature-enablement-and-rollback)
+  - [Rollout, Upgrade and Rollback Planning](#rollout-upgrade-and-rollback-planning)
+  - [Monitoring Requirements](#monitoring-requirements)
+  - [Dependencies](#dependencies)
+  - [Scalability](#scalability)
+  - [Troubleshooting](#troubleshooting)
+- [Implementation History](#implementation-history)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+<!-- /toc -->
+
+## Release Signoff Checklist
+
+Items marked with (R) are required *prior to targeting to a milestone / release*.
+
+- [ ] (R) Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
+- [ ] (R) KEP approvers have approved the KEP status as `implementable`
+- [ ] (R) Design details are appropriately documented
+- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input (including test refactors)
+- [ ] (R) Graduation criteria is in place
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+This KEP proposes a feature to protect Secrets/ConfigMaps while it is in use. Currently, users can delete a secret that is being used by other resources, like Pods, PVs, and VolumeSnapshots. This may have negative impact on the resouces using the secret and it may result in data loss.
+
+For example, one of the worst case scenarios is deletion of a Node Stage Secret for CSI volumes while it is still in-use.
+If it happens, a CSI driver for the volume can't unstage the volume due to the lack of the Secret,
```

Comment:
even if the secret API object can be kept from being deleted, there are a variety of other situations that could cause this:
* the contents of the secret are modified and no longer valid credentials
* the credentials change and what is in the secret is no longer valid

I would expect cluster-level administrative objects like secrets required to unmount/unstage persistent volumes to be placed in administrative namespaces, not in a user-accessible namespace alongside the PVC

Reply:
> Even if the secret API object can be kept from being deleted, there are a variety of other situations that could cause this:
> 
> - the contents of the secret are modified by the user and no longer valid credentials
> - the credentials change and what is in the secret is no longer valid
> Being able to safely tear down volumes in those situations seems important as well.

I agree that such a situation can still happen.

For the first bullet, immutable secret will be the solution.
For the second bullet, we won't have a general solution in the k8s cluster side, technically. It will only happen if backend is directly modified by administrators, and in such a case, administrators should modify the secret by themselves (Because of it, I think that such changes won't be recommended.).

> Separately, I would expect cluster-level administrative objects like secrets required to unmount/unstage persistent volumes to be placed in administrative namespaces, not in a user-accessible namespace alongside the PVC

Changing the [specification for CSI secret](https://kubernetes-csi.github.io/docs/secrets-and-credentials.html) will be one way to solve this kind of issue. However, it will result in prohibiting [per volume secrets use cases](https://kubernetes-csi.github.io/docs/secrets-and-credentials-storage-class.html#per-volume-secrets).

We will need to be careful when choosing this direction.
Also, this feature is still needed until the spec is completely transitioned to such a way.

Reply:
> For the first bullet, immutable secret will be the solution.

Added below to non-goal to make it clear.
"- Protect Secrets/ConfigMaps from being __updated__ while they are in use (Immutable Secrets/ConfigMaps will solve it)"